---
title: "Introduction to Spline-DV"
author: 
- name: Shreyan Gupta
  affiliation: Veterinary Integrative Biosciences, Texas A&M University
- name: Victoria Gatlin
  affiliation: Veterinary Integrative Biosciences, Texas A&M University
- name: Selim Romero
  affiliation: Department of Nutrition, Texas A&M University
- name: James J Cai
  affiliation: Veterinary Integrative Biosciences, Texas A&M University
abstract: "This tutorial shows how to use **Spline-DV** for Differential Variability
  Analysis across two single-cell RNA seq samples and **Spline-HVG** for highly variable
  feature selection"
output:
  BiocStyle::html_document:
    toc: true
    number_sections: false
  html_notebook:
    toc: true
  pdf_document:
    toc: true
    
vignette: >
  %\VignetteIndexEntry{Introduction to Spline-DV}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo=TRUE)
```

# Why to use SplineDV? 
## Beyond Averages
One of the most intuitive ways to evaluate a gene expression change is using Differential Expression (DE) analysis. Traditionally, DE analysis focuses on identifying genes that are up- or down-regulated (increased or decreased expression) between conditions, typically employing a basic mean-difference approach. We propose a paradigm shift that acknowledges the central role of gene expression variability in cellular function and challenges the current dominance of mean-based DE analysis in single-cell studies. We suggest that scRNA-seq data analysis should embrace the role of inherent gene expression variability in defining cellular function and move beyond mean-based approaches. 

# Installation
```{r Installation, eval=FALSE}
BiocManager::install("Xenon8778/SplineDV")
```

# Overview of the SplineDV workflow 
## Input - Two expression matrix 
The expression matrix should be formatted with the genes/features as rows and cells as columns.
The two input matrices for DV_splinefit must be stored as two separate count expression matrices or SingleCellExperiment objects.

## Loading scRNAseq count example data
The example data is borrowed from an experimental *Nkx2-1* Gene knockout scRNA-seq study by Liebler *et al.* ^1^

```{r loadingExampledata}
# Load Data
load(system.file("extdata", "WT_count.rda", package = "SplineDV")) # WT Sample
load(system.file("extdata", "KO_count.rda", package = "SplineDV")) # KO Sample
```

## Running Spline-DV
For the analysis, the test data (X) is always use in contrast with the control data (Y). We use smaller QC parameters for the small example data sets. **We recommend using the default QC parameters for large data sets.**
```{r runSplineDV, results='hide'}
library(SplineDV)
DV_res <- DV_splinefit(X = KO_count, Y = WT_count, ncells = 3, ncounts = 200)
```

```{r showResults, echo=FALSE}
head(DV_res)
```

## Visualize Gene Expression statistics

```{r plotDV, out.width="75%", out.height="75%", fig.format='png'}
library(plotly)
fig <- DV_plot(DV_res)
fig 
```

 
# Highly Variable Genes (HVGs) using Spline-HVG 
## Input - scRNA-seq Expression matrix 
```{r LoadHVGData}
# Loading Data
load(system.file("extdata", "WT_count.rda", package = "SplineDV")) # WT Sample
```
## Running Spline-HVG
```{r runSplineHVG, results='hide'}
HVG_res <- HVG_splinefit(WT_count, nHVGs = 20, ncells = 3, ncounts = 200)
```
```{r showHVGResults, echo=FALSE}
head(HVG_res)
```
```{r showHVGList}
# Extracting HVG Gene list
HVG_list <- rownames(HVG_res)[HVG_res$HVG == TRUE]
head(HVG_list)
```
## Visualize Highly Variable Genes
```{r plotHVG, out.width="75%", out.height="75%"}
fig <- HVG_plot(HVG_res)
fig
```

# Appendix
## Citing SplineDV
```{r "citation"}
citation("SplineDV")
```
## References
1. Liebler JM, Marconett CN, Juul N, et al. Combinations of differentiation markers distinguish subpopulations of alveolar epithelial cells in adult lung. Am J Physiol Lung Cell Mol Physiol. 2016;310(2):L114-L120. doi:10.1152/ajplung.00337.2015

# sessionInfo
This is the output of `sessionInfo()` on the system on which this document was compiled:
```{r}
date()
sessionInfo()
```
