---
title: "Introduction to Spline-DV"
author: 
- name: Shreyan Gupta
  affiliation: Veterinary Integrative Biosciences, Texas A&M University
- name: Victoria Gatlin
  affiliation: Veterinary Integrative Biosciences, Texas A&M University
- name: Selim Romero
  affiliation: Department of Nutrition, Texas A&M University
- name: James J Cai
  affiliation: Veterinary Integrative Biosciences, Texas A&M University
abstract: "This tutorial shows how to use **Spline-DV** for Differential Variability
  Analysis across two single-cell RNA seq samples and **Spline-HVG** for highly variable
  feature selection"
output:
  BiocStyle::html_document:
    toc: true
    number_sections: false
  html_notebook:
    toc: true
  pdf_document:
    toc: true
    
vignette: >
  %\VignetteIndexEntry{Introduction to Spline-DV}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo=TRUE)
```

# Why to use SplineDV? 
## Beyond Averages
One of the most intuitive ways to evaluate a gene expression change is using Differential Expression (DE) analysis. Traditionally, DE analysis focuses on identifying genes that are up- or down-regulated (increased or decreased expression) between conditions, typically employing a basic mean-difference approach. We propose a paradigm shift that acknowledges the central role of gene expression variability in cellular function and challenges the current dominance of mean-based DE analysis in single-cell studies. We suggest that scRNA-seq data analysis should embrace the role of inherent gene expression variability in defining cellular function and move beyond mean-based approaches. 

# Installation
```{r Installation, eval=FALSE}
BiocManager::install("Xenon8778/SplineDV")
```

# Overview of the SplineDV workflow 
## Input - Two expression matrix 
The expression matrix should be formatted with the genes/features as rows and cells as columns.
The two input matrices for splineDV must be stored as two separate count expression matrices or SingleCellExperiment objects.

## Loading scRNAseq count example data
The example data is borrowed from an experimental *Nkx2-1* Gene knockout scRNA-seq study by Liebler *et al.* ^1^

```{r loadingExampledata}
# Load Data
WTcount <- SplineDV::WTcount
KOcount <- SplineDV::KOcount
```

## Running Spline-DV
For the analysis, the test data (X) is always use in contrast with the control data (Y). We use smaller QC parameters for the small example data sets to preserve enough cells and genes. **We recommend using the default QC parameters for large data sets.**
```{r runSplineDV, results='hide'}
library(SplineDV)
dvRes <- splineDV(X = KOcount, Y = WTcount, ncells = 3, ncounts = 200)
```

```{r showResults, echo=FALSE}
head(dvRes)
```

### The splineDV result data.frame contains the following data:
- **genes** - Contains the name of the gene.
- **mu1** - log1p(mean) expression of gene in test sample.
- **mu2** - log1p(mean) expression of gene in control sample.
- **CV1** - log1p(CV) expression of gene in test sample.
- **CV2** - log1p(CV) expression of gene in control sample.
- **drop1** - Dropout rate of gene in test sample.
- **drop2** - Dropout rate of gene in control sample.
- **dist1** - Distance of real gene statistics from nearest point on spline in test sample.
- **dist2** - Distance of real gene statistics from nearest point on spline in control sample.
- **X_splinex**, **X_spliney**, **X_splinez** - (x,y,z) coordinates of nearest point on spline in test sample.
- **Y_splinex**, **Y_spliney**, **Y_splinez** - (x,y,z) coordinates of nearest point on spline in control sample.
- **vectorDist** - Vector distance between distance vectors between real expression and spline for test and control samples.
- **Direction** - Direction of variability change
- **Pval** - P-value computed from gaussian distribution.

## Visualize Gene Expression statistics
```{r Load Plotly, message=FALSE}
library(plotly)
```

A 3D scatter plot with computed spline of each condition in respective color and the distance vectors of genes from respective spline.

```{r plotDV, out.width="75%", out.height="75%", fig.format='png'}
fig <- DVPlot(dvRes)
fig 
```

 
# Highly Variable Genes (HVGs) using Spline-HVG 
## Input - scRNA-seq Expression matrix 
```{r LoadHVGData}
# Loading Data
WTcount <- SplineDV::WTcount
```
## Running Spline-HVG
We use smaller QC parameters for the small example data sets to preserve enough cells and genes. **We recommend using the default QC parameters for large data sets.**
```{r runSplineHVG, results='hide'}
HVGRes <- splineHVG(WTcount, nHVGs = 20, ncells = 3, ncounts = 200)
```
```{r showHVGResults, echo=FALSE}
head(HVGRes)
```
```{r showHVGList}
# Extracting HVG Gene list
HVGList <- rownames(HVGRes)[HVGRes$HVG == TRUE]
head(HVGList)
```

### The splineHVG result data.frame contains the following data:

- **Means** - mean expression of gene.
- **logMean** - log1p(mean) expression of gene.
- **CV** - log1p(CV) expression of gene.
- **logCV** - log1p(CV) expression of gene.
- **Dropout** - Dropout rate of gene.
- **Distance** - Distance of real gene statistics from nearest point on spline.
- **nearidx** - Index of nearest point on spline.
- **dvecx**, **dvecy**, **dvecy** - Distances on (x,y,z) axis of real gene statistics from nearest point on spline.
- **splinex**, **spliney**, **splinez** - (x,y,z) coordinates of nearest point on spline.
- **HVG** - Boolean value indicating whether the gene is an HVG.

## Visualize Highly Variable Genes
A 3D scatter plot with computed spline in black. Highly variable genes are highlighted in red, non-HVGs are grey.
```{r plotHVG, out.width="75%", out.height="75%"}
fig <- HVGPlot(HVGRes)
fig
```

# Appendix
## Citing SplineDV
```{r "citation"}
citation("SplineDV")
```
## References
1. Liebler JM, Marconett CN, Juul N, et al. Combinations of differentiation markers distinguish subpopulations of alveolar epithelial cells in adult lung. Am J Physiol Lung Cell Mol Physiol. 2016;310(2):L114-L120. doi:10.1152/ajplung.00337.2015

# sessionInfo
This is the output of `sessionInfo()` on the system on which this document was compiled:
```{r}
date()
sessionInfo()
```
